### ISSUE TITLE
Scrolling “too long” command doesn't work properly

### ISSUE THREAD

**Issue by EdGue42** [fish-shell/fish-shell#6014]
I am running fish, version 3.0.2 on MacOs (latest greatest 10.14.6). The problem exists on the built-in Apple iterm, but also on the iterm2 (TERM says: xterm-256color).

When using _sh -c 'env HOME=$(mktemp -d) fish'_ and then I can **not** repro my problem!

But anyway, the problem is this.

Assume I want to invoke a tool with a lot of options, like:

> $ somescript --option1 --option2 --option3 --option4 --option5 whatever even more stuff
> But the width of the terminal doesn't allow to have that thing on a "single line".

That alone isn't a problem, but with fish, there is a problem with scrolling.
When I enter that extra-wide command, at some point, there will be a line break:

So, assume I entered the command, made a mistake, and now want to remove --option4.

- When I dit nit hit enter yet: I simply hit "arrow left" to move the caret to the place I need to edit
- When I did hit enter: I use "arrow up" to get to that command in the history, to then use "arrow left" to scroll within the command. 

In both cases, something weird happens: everything is fine while the cursor is within the second line. But when the cursor moves to the first line ... after 3 to 5 more "arrow left" strokes, the cursor moves UP another line. Then it sits above the first line of the command. There is also a vertical jump of a few characters.

In other words: as soon as a command is longer than the width of the terminal, I am unable to scroll into the first line reliably. The cursor shows up somewhere, and it is really hard to guess where it really sits at any moment. Which makes it almost impossible to edit anything in that first line.

Addendum: when using the "pure fish" behavior is this: as soon as I reach "end of line" for the first time, the command automatically "moves" to a new line ... and scrolling forth and back seems to work all the time.

**Comment by zanchey** [comment]
This typically happens when there are characters either in the command line or in your prompt with a width that fish and the terminal disagree on - can you share a screenshot?

**Comment by EdGue42** [comment]
Me typing the command: 

![image](https://user-images.githubusercontent.com/53468311/62114221-7ba95e00-b2b6-11e9-9eb3-50c644223196.png)

Me scrolling to the left:

![image](https://user-images.githubusercontent.com/53468311/62114272-98459600-b2b6-11e9-829e-cb042683759a.png)

You can notice that the white caret is now one line above the line I am scrolling. And it also has a horizontal offset. 

Does that help (and note that I edited my initial input to be more clear).

**Comment by faho** [comment]
@EdGue42 Your prompt contains a bunch of non-ascii characters. Please replace it with

```fish
function fish_prompt
    echo $PWD '>'
end
```

and try again.

**Comment by EdGue42** [comment]
@faho I did so, and yes, that makes that problem go away.

**But**: I simply selected that prompt via the fish_config tool. I picked the "AcidHub" prompt from there, as that includes GIT repository information.

And I picked that prompt because it shows me the information I need. 
So, yes, your suggestion fixes the the issue, but it doesn't really help me. 

In other words: I understand that the "broken" prompt is causing the scrolling issue. But well, that "broken" prompt was suggested to me by fish_config. 

So: is there a way to fix that "AcidHub"? Maybe first locally for me, but also in a way that other people do not run into the same problem?

**Comment by faho** [comment]
>So, maybe, just maybe, the problem is that the AcidHub template prompt contains stuff like:
set git_status "$git_status"(set_color red)<E2><AC><86>

No, the problem *definitely* is that your terminal renders these characters in a different width than fish expects. The solution is to either adjust fish's expectation or the terminal's rendering.

You can try setting $fish_emoji_width or $fish_ambiguous_width to 1 or 2, in the four possible combinations. I would imagine setting $fish_emoji_width to 2 should do it (and would be auto-detected in fish 3.1).

**Comment by EdGue42** [comment]
I think I tried all 4 combos: e width 1/2 and a width 1/2.
Alas, it didn't seem to help.

Where: I did the set ... command in one iterm2 fish tab, and then started another tab. To see that scrolling left for long commands still gives me the "jumping" caret. 

Is there a simple way how I could create a ascii-only version of AcidHub?

**Comment by faho** [comment]
>Where: I did the set ... command in one iterm2 fish tab, and then started another tab

NOoooooo! The other tab won't get the value. Set it in that tab and then use that tab, or set it in config.fish.

**Comment by EdGue42** [comment]
Unfortunately, nothing seems to help. Some of the settings seem to help when typing the initial command ... but latest when I go "arrow up" first, the editor caret keeps jumping around. 

Well, to be precise: 

- I put _set -x fish_emoji_width 2_ into my fish.config
- when I open a new terminal, and go "arrow up" to then scroll around in some existing command ... that works fine. The first time.
- when I now hit enter, and do the very same, "arrow up" ... the left scrolling gets jumpy. 

( I am still open for suggestions how to rework the AcidHub to not include emojis and stuff ;-)

**Comment by faho** [comment]
>Some of the settings seem to help when typing the initial command

Keep those, then look at the commands that are causing the issue. Most likely they have other characters that are causing other issues.

The main problem is this:

- If your terminal is wrong, there's nothing we can do (and tbh this appears to be quite likely. I've checked all of the non-ascii chars in acidhub, and they should all be classified as "narrow")
- If your terminal displays different ambiguous characters with different widths, there's nothing we can do.

Also since these are git characters, you might need to try all of them. Alternatively, execute

```fish
printf '%s\n' mm ✚ ✖ ✱ ➜ ═ ≠ ❰ ❙ ❱ ✔ ✘ ≻
```

and see how all of these compare to the "mm". 

>( I am still open for suggestions how to rework the AcidHub to not include emojis and stuff ;-)

`funced fish_prompt`. When you're happy with it `funcsave fish_prompt`.

**Comment by EdGue42** [comment]
I am using always the same command right now, a simple "echo word1 word2 ...". 
And I am running it directly from my home directory, so it really only gives my user name + ~ . And well, the green "checked" from git I guess.

**Comment by faho** [comment]
>Seems like those trailing ✔ ✘ ≻ give issues, at least when I did the copy/paste?

Yes, which means your terminal is wrong. "≻" is U+227B, "✔" is U+2714, ✘ is U+2718.

Consulting [EastAsianWidth.txt](https://www.unicode.org/Public/UNIDATA/EastAsianWidth.txt) (the relevant standard):

>2270..2281;N     # Sm    [18] NEITHER LESS-THAN NOR EQUAL TO..DOES NOT SUCCEED
> [...]
>270C..2727;N     # So    [28] VICTORY HAND..WHITE FOUR POINTED STAR

Where that "N" means "narrow" i.e. width-of-1.

Sorry, there is no way for us to support this, so on this system you'll just have to pick a different prompt.

**Comment by EdGue42** [comment]
Are you aware of any terminal software for MacOs that does **not** have this problem?

**Comment by faho** [comment]
No, I am not specifically aware of any terminal software for macOS.

I would assume it's an issue in the OS-provided text rendering, so most terminals are likely to have it.

**Comment by EdGue42** [comment]
Frustrating, but obviously not your responsibility.

Thank you very much for the quick comebacks, and the time you spent to help with this.
